### ISSUE TITLE
Insecure registry test failure: FetchAuthToken

### ISSUE THREAD

**Issue by mhagen-vmware** [vmware/vic#5473]
Failed these jobs:
https://ci.vcna.io/vmware/vic/11504
https://ci.vcna.io/vmware/vic/11495

```
Jun 13 2017 22:20:28.540Z DEBUG Trying https scheme for &imagec.Options{Reference:(*reference.taggedRef)(0xc42060ff00), Registry:"192.168.31.236", Image:"test/busybox", Tag:"latest", Destination:"/tmp", Host:"127.0.0.1:2377", Storename:"564dd154-83f0-7b14-bd03-3e9e2ab1a7a4", Username:"<user>", Password:"<password>", Token:(*fetcher.Token)(nil), Timeout:3600000000000, Outstream:(*ioutils.WriteFlusher)(0xc420180550), InsecureSkipVerify:false, InsecureAllowHTTP:true, ImageManifestSchema1:(*imagec.Manifest)(nil), ImageManifestSchema2:(*schema2.DeserializedManifest)(nil), ManifestDigest:"", RegistryCAs:(*x509.CertPool)(0xc42045e3f0)}
Jun 13 2017 22:20:28.540Z DEBUG URL: https://192.168.31.236/v2/
Jun 13 2017 22:20:28.540Z DEBUG [BEGIN] [github.com/vmware/vic/pkg/fetcher.(*URLFetcher).Head:361] https://192.168.31.236/v2/
Jun 13 2017 22:20:28.586Z DEBUG [ END ] [github.com/vmware/vic/pkg/fetcher.(*URLFetcher).Head:361] [45.84525ms] https://192.168.31.236/v2/
Jun 13 2017 22:20:28.588Z DEBUG [ END ] [github.com/vmware/vic/lib/imagec.LearnRegistryURL:75] [47.30936ms] 192.168.31.236
Jun 13 2017 22:20:28.589Z DEBUG [BEGIN] [github.com/vmware/vic/lib/imagec.LearnAuthURL:100] 192.168.31.236/test/busybox:latest
Jun 13 2017 22:20:28.589Z DEBUG URL: https://192.168.31.236/v2/test/busybox/manifests/latest
Jun 13 2017 22:20:28.589Z DEBUG [BEGIN] [github.com/vmware/vic/pkg/fetcher.(*URLFetcher).Head:361] https://192.168.31.236/v2/test/busybox/manifests/latest
Jun 13 2017 22:20:28.613Z DEBUG [ END ] [github.com/vmware/vic/pkg/fetcher.(*URLFetcher).Head:361] [23.910073ms] https://192.168.31.236/v2/test/busybox/manifests/latest
Jun 13 2017 22:20:28.613Z DEBUG [ END ] [github.com/vmware/vic/lib/imagec.LearnAuthURL:100] [23.973389ms] 192.168.31.236/test/busybox:latest
Jun 13 2017 22:20:28.613Z DEBUG [BEGIN] [github.com/vmware/vic/lib/imagec.FetchToken:144] https://localhost.localdom.ci.vmware.run/service/token?scope=repository%3Atest%2Fbusybox%3Apull&service=token-service
Jun 13 2017 22:20:28.613Z DEBUG URL: https://localhost.localdom.ci.vmware.run/service/token?scope=repository%3Atest%2Fbusybox%3Apull&service=token-service
Jun 13 2017 22:20:28.613Z DEBUG [BEGIN] [github.com/vmware/vic/pkg/fetcher.(*URLFetcher).FetchAuthToken:198] https://localhost.localdom.ci.vmware.run/service/token?scope=repository%3Atest%2Fbusybox%3Apull&service=token-service
Jun 13 2017 22:20:28.613Z DEBUG [BEGIN] [github.com/vmware/vic/pkg/fetcher.(*URLFetcher).Fetch:124] https://localhost.localdom.ci.vmware.run/service/token?scope=repository%3Atest%2Fbusybox%3Apull&service=token-service
Jun 13 2017 22:20:28.613Z DEBUG Setting BasicAuth: <user>
Jun 13 2017 22:20:28.613Z DEBUG Setting user-agent to vic/v1.1.0-rc3
Jun 13 2017 22:20:28.626Z ERROR Fetch (https://localhost.localdom.ci.vmware.run/service/token?scope=repository%3Atest%2Fbusybox%3Apull&service=token-service) to string error: Get https://localhost.localdom.ci.vmware.run/service/token?scope=repository%3Atest%2Fbusybox%3Apull&service=token-service: dial tcp 192.168.31.160:443: getsockopt: connection refused
Jun 13 2017 22:20:28.626Z DEBUG Download failed, retrying: Get https://localhost.localdom.ci.vmware.run/service/token?scope=repository%3Atest%2Fbusybox%3Apull&service=token-service: dial tcp 192.168.31.160:443: getsockopt: connection refused
Jun 13 2017 22:20:33.626Z DEBUG Setting BasicAuth: <user>
Jun 13 2017 22:20:33.626Z DEBUG Setting user-agent to vic/v1.1.0-rc3
Jun 13 2017 22:20:33.628Z ERROR Fetch (https://localhost.localdom.ci.vmware.run/service/token?scope=repository%3Atest%2Fbusybox%3Apull&service=token-service) to string error: Get https://localhost.localdom.ci.vmware.run/service/token?scope=repository%3Atest%2Fbusybox%3Apull&service=token-service: dial tcp 192.168.31.160:443: getsockopt: connection refused
Jun 13 2017 22:20:33.628Z DEBUG Download failed, retrying: Get https://localhost.localdom.ci.vmware.run/service/token?scope=repository%3Atest%2Fbusybox%3Apull&service=token-service: dial tcp 192.168.31.160:443: getsockopt: connection refused
Jun 13 2017 22:20:43.629Z DEBUG Setting BasicAuth: <user>
Jun 13 2017 22:20:43.629Z DEBUG Setting user-agent to vic/v1.1.0-rc3
Jun 13 2017 22:20:43.633Z ERROR Fetch (https://localhost.localdom.ci.vmware.run/service/token?scope=repository%3Atest%2Fbusybox%3Apull&service=token-service) to string error: Get https://localhost.localdom.ci.vmware.run/service/token?scope=repository%3Atest%2Fbusybox%3Apull&service=token-service: dial tcp 192.168.31.160:443: getsockopt: connection refused
Jun 13 2017 22:20:43.633Z DEBUG Download failed, retrying: Get https://localhost.localdom.ci.vmware.run/service/token?scope=repository%3Atest%2Fbusybox%3Apull&service=token-service: dial tcp 192.168.31.160:443: getsockopt: connection refused
Jun 13 2017 22:20:58.633Z DEBUG Setting BasicAuth: <user>
Jun 13 2017 22:20:58.633Z DEBUG Setting user-agent to vic/v1.1.0-rc3
Jun 13 2017 22:21:16.670Z ERROR Fetch (https://localhost.localdom.ci.vmware.run/service/token?scope=repository%3Atest%2Fbusybox%3Apull&service=token-service) to string error: Get https://localhost.localdom.ci.vmware.run/service/token?scope=repository%3Atest%2Fbusybox%3Apull&service=token-service: dial tcp 192.168.31.160:443: getsockopt: no route to host
Jun 13 2017 22:21:16.670Z DEBUG Download failed, retrying: Get https://localhost.localdom.ci.vmware.run/service/token?scope=repository%3Atest%2Fbusybox%3Apull&service=token-service: dial tcp 192.168.31.160:443: getsockopt: no route to host
Jun 13 2017 22:21:36.670Z DEBUG Setting BasicAuth: <user>
Jun 13 2017 22:21:36.670Z DEBUG Setting user-agent to vic/v1.1.0-rc3
Jun 13 2017 22:21:39.670Z ERROR Fetch (https://localhost.localdom.ci.vmware.run/service/token?scope=repository%3Atest%2Fbusybox%3Apull&service=token-service) to string error: Get https://localhost.localdom.ci.vmware.run/service/token?scope=repository%3Atest%2Fbusybox%3Apull&service=token-service: dial tcp 192.168.31.160:443: getsockopt: no route to host
Jun 13 2017 22:21:39.670Z DEBUG Hit max download attempts. Download failed: Get https://localhost.localdom.ci.vmware.run/service/token?scope=repository%3Atest%2Fbusybox%3Apull&service=token-service: dial tcp 192.168.31.160:443: getsockopt: no route to host
Jun 13 2017 22:21:39.670Z DEBUG [ END ] [github.com/vmware/vic/pkg/fetcher.(*URLFetcher).Fetch:124] [1m11.056943456s] https://localhost.localdom.ci.vmware.run/service/token?scope=repository%3Atest%2Fbusybox%3Apull&service=token-service
Jun 13 2017 22:21:39.670Z ERROR Download failed: Get https://localhost.localdom.ci.vmware.run/service/token?scope=repository%3Atest%2Fbusybox%3Apull&service=token-service: dial tcp 192.168.31.160:443: getsockopt: no route to host
Jun 13 2017 22:21:39.670Z DEBUG [ END ] [github.com/vmware/vic/pkg/fetcher.(*URLFetcher).FetchAuthToken:198] [1m11.057005817s] https://localhost.localdom.ci.vmware.run/service/token?scope=repository%3Atest%2Fbusybox%3Apull&service=token-service
Jun 13 2017 22:21:39.670Z ERROR FetchToken (https://localhost.localdom.ci.vmware.run/service/token?scope=repository%3Atest%2Fbusybox%3Apull&service=token-service) failed: Get https://localhost.localdom.ci.vmware.run/service/token?scope=repository%3Atest%2Fbusybox%3Apull&service=token-service: dial tcp 192.168.31.160:443: getsockopt: no route to host
Jun 13 2017 22:21:39.670Z DEBUG [ END ] [github.com/vmware/vic/lib/imagec.FetchToken:144] [1m11.05707475s] https://localhost.localdom.ci.vmware.run/service/token?scope=repository%3Atest%2Fbusybox%3Apull&service=token-service
Jun 13 2017 22:21:39.670Z ERROR Failed to fetch OAuth token: FetchToken (https://localhost.localdom.ci.vmware.run/service/token?scope=repository%3Atest%2Fbusybox%3Apull&service=token-service) failed: Get https://localhost.localdom.ci.vmware.run/service/token?scope=repository%3Atest%2Fbusybox%3Apull&service=token-service: dial tcp 192.168.31.160:443: getsockopt: no route to host
Jun 13 2017 22:21:39.670Z DEBUG [ END ] [github.com/vmware/vic/lib/apiservers/engine/backends.(*Image).PullImage:325] [1m11.143906612s] 
Jun 13 2017 22:21:39.670Z ERROR Handler for POST /v1.23/images/create returned error: FetchToken (https://localhost.localdom.ci.vmware.run/service/token?scope=repository%3Atest%2Fbusybox%3Apull&service=token-service) failed: Get https://localhost.localdom.ci.vmware.run/service/token?scope=repository%3Atest%2Fbusybox%3Apull&service=token-service: dial tcp 192.168.31.160:443: getsockopt: no route to host
```

So, for some reason it looks like midway through this process we switched from the valid IP address of 192.168.31.236  to 192.168.31.160.  On a different successful run, it did not switch IPs.

**Comment by mhagen-vmware** [comment]
On a successful run, we stick with the valid IP addess of 208:
```
Jun 13 2017 22:10:29.554Z DEBUG Trying https scheme for &imagec.Options{Reference:(*reference.taggedRef)(0xc4208ad9c0), Registry:"192.168.31.208", Image:"test/busybox", Tag:"latest", Destination:"/tmp", Host:"127.0.0.1:2377", Storename:"564d1c96-aa3c-8b7c-8071-23961409c2b6", Username:"admin", Password:"Og0928V7WIUj", Token:(*fetcher.Token)(nil), Timeout:3600000000000, Outstream:(*ioutils.WriteFlusher)(0xc420214af0), InsecureSkipVerify:false, InsecureAllowHTTP:true, ImageManifestSchema1:(*imagec.Manifest)(nil), ImageManifestSchema2:(*schema2.DeserializedManifest)(nil), ManifestDigest:"", RegistryCAs:(*x509.CertPool)(0xc4203ae9c0)}
Jun 13 2017 22:10:29.554Z DEBUG URL: https://192.168.31.208/v2/
Jun 13 2017 22:10:29.554Z DEBUG [BEGIN] [github.com/vmware/vic/pkg/fetcher.(*URLFetcher).Head:361] https://192.168.31.208/v2/
Jun 13 2017 22:10:29.605Z DEBUG [ END ] [github.com/vmware/vic/pkg/fetcher.(*URLFetcher).Head:361] [51.24219ms] https://192.168.31.208/v2/
Jun 13 2017 22:10:29.609Z DEBUG [ END ] [github.com/vmware/vic/lib/imagec.LearnRegistryURL:75] [55.107183ms] 192.168.31.208
Jun 13 2017 22:10:29.609Z DEBUG [BEGIN] [github.com/vmware/vic/lib/imagec.LearnAuthURL:100] 192.168.31.208/test/busybox:latest
Jun 13 2017 22:10:29.609Z DEBUG URL: https://192.168.31.208/v2/test/busybox/manifests/latest
Jun 13 2017 22:10:29.609Z DEBUG [BEGIN] [github.com/vmware/vic/pkg/fetcher.(*URLFetcher).Head:361] https://192.168.31.208/v2/test/busybox/manifests/latest
Jun 13 2017 22:10:29.633Z DEBUG [ END ] [github.com/vmware/vic/pkg/fetcher.(*URLFetcher).Head:361] [24.320096ms] https://192.168.31.208/v2/test/busybox/manifests/latest
Jun 13 2017 22:10:29.633Z DEBUG [ END ] [github.com/vmware/vic/lib/imagec.LearnAuthURL:100] [24.377971ms] 192.168.31.208/test/busybox:latest
Jun 13 2017 22:10:29.633Z DEBUG [BEGIN] [github.com/vmware/vic/lib/imagec.FetchToken:144] https://localhost.localdom.ci.vmware.run/service/token?scope=repository%3Atest%2Fbusybox%3Apull&service=token-service
Jun 13 2017 22:10:29.633Z DEBUG URL: https://localhost.localdom.ci.vmware.run/service/token?scope=repository%3Atest%2Fbusybox%3Apull&service=token-service
Jun 13 2017 22:10:29.633Z DEBUG [BEGIN] [github.com/vmware/vic/pkg/fetcher.(*URLFetcher).FetchAuthToken:198] https://localhost.localdom.ci.vmware.run/service/token?scope=repository%3Atest%2Fbusybox%3Apull&service=token-service
Jun 13 2017 22:10:29.633Z DEBUG [BEGIN] [github.com/vmware/vic/pkg/fetcher.(*URLFetcher).Fetch:124] https://localhost.localdom.ci.vmware.run/service/token?scope=repository%3Atest%2Fbusybox%3Apull&service=token-service
Jun 13 2017 22:10:29.633Z DEBUG Setting BasicAuth: admin
Jun 13 2017 22:10:29.633Z DEBUG Setting user-agent to vic/v1.1.0-rc3
Jun 13 2017 22:10:29.704Z DEBUG URLFetcher.fetch() - &http.bodyEOFSignal{body:(*http.body)(0xc4208c1dc0), mu:sync.Mutex{state:0, sema:0x0}, closed:false, rerr:error(nil), fn:(func(error) error)(0x6fee80), earlyCloseFn:(func() error)(0x6fee10)}, http.Header{"Server":[]string{"nginx/1.11.5"}, "Date":[]string{"Tue, 13 Jun 2017 22:10:30 GMT"}, "Content-Type":[]string{"application/json; charset=utf-8"}, "Content-Length":[]string{"1194"}, "Connection":[]string{"keep-alive"}, "Set-Cookie":[]string{"beegosessionID=090f59cf3695ac7be8dabbbb122a552a; Path=/; HttpOnly"}}
Jun 13 2017 22:10:29.704Z DEBUG [ END ] [github.com/vmware/vic/pkg/fetcher.(*URLFetcher).Fetch:124] [70.62998ms] https://localhost.localdom.ci.vmware.run/service/token?scope=repository%3Atest%2Fbusybox%3Apull&service=token-service
Jun 13 2017 22:10:29.704Z DEBUG [ END ] [github.com/vmware/vic/pkg/fetcher.(*URLFetcher).FetchAuthToken:198] [70.713459ms] https://localhost.localdom.ci.vmware.run/service/token?scope=repository%3Atest%2Fbusybox%3Apull&service=token-service
Jun 13 2017 22:10:29.704Z DEBUG [ END ] [github.com/vmware/vic/lib/imagec.FetchToken:144] [70.753472ms] https://localhost.localdom.ci.vmware.run/service/token?scope=repository%3Atest%2Fbusybox%3Apull&service=token-service
Jun 13 2017 22:10:29.704Z DEBUG [BEGIN] [github.com/vmware/vic/lib/imagec.FetchImageManifest:313] 192.168.31.208/test/busybox:latest
Jun 13 2017 22:10:29.704Z DEBUG URL: https://192.168.31.208/v2/test/busybox/manifests/latest
Jun 13 2017 22:10:29.704Z DEBUG [BEGIN] [github.com/vmware/vic/pkg/fetcher.(*URLFetcher).Fetch:124] https://192.168.31.208/v2/test/busybox/manifests/latest
Jun 13 2017 22:10:29.704Z DEBUG Setting BasicAuth: admin
Jun 13 2017 22:10:29.704Z DEBUG Setting user-agent to vic/v1.1.0-rc3
Jun 13 2017 22:10:29.744Z DEBUG URLFetcher.fetch() - &http.bodyEOFSignal{body:(*http.body)(0xc4201fab40), mu:sync.Mutex{state:0, sema:0x0}, closed:false, rerr:error(nil), fn:(func(error) error)(0x6fee80), earlyCloseFn:(func() error)(0x6fee10)}, http.Header{"Content-Type":[]string{"application/vnd.docker.distribution.manifest.v1+prettyjws"}, "Etag":[]string{"\"sha256:6c309d7c5f93f8c2e23f385e98eade82c1e6333b6cc36562d81f44a86e3c16d9\""}, "Server":[]string{"nginx/1.11.5"}, "Date":[]string{"Tue, 13 Jun 2017 22:10:30 GMT"}, "Content-Length":[]string{"2786"}, "Connection":[]string{"keep-alive"}, "Docker-Content-Digest":[]string{"sha256:6c309d7c5f93f8c2e23f385e98eade82c1e6333b6cc36562d81f44a86e3c16d9"}, "Docker-Distribution-Api-Version":[]string{"registry/2.0"}}
Jun 13 2017 22:10:29.744Z DEBUG [ END ] [github.com/vmware/vic/pkg/fetcher.(*URLFetcher).Fetch:124] [39.996284ms] https://192.168.31.208/v2/test/busybox/manifests/latest
Jun 13 2017 22:10:29.744Z DEBUG Canonical Bytes: 2141
Jun 13 2017 22:10:29.744Z DEBUG Manifest Digest: sha256:6c309d7c5f93f8c2e23f385e98eade82c1e6333b6cc36562d81f44a86e3c16d9
Jun 13 2017 22:10:29.744Z DEBUG [ END ] [github.com/vmware/vic/lib/imagec.FetchImageManifest:313] [40.375292ms] 192.168.31.208/test/busybox:latest
Jun 13 2017 22:10:29.744Z DEBUG [BEGIN] [github.com/vmware/vic/lib/imagec.FetchImageManifest:313] 192.168.31.208/test/busybox:latest
Jun 13 2017 22:10:29.744Z DEBUG URL: https://192.168.31.208/v2/test/busybox/manifests/latest
Jun 13 2017 22:10:29.744Z DEBUG [BEGIN] [github.com/vmware/vic/pkg/fetcher.(*URLFetcher).Fetch:124] https://192.168.31.208/v2/test/busybox/manifests/latest
Jun 13 2017 22:10:29.744Z DEBUG Setting BasicAuth: admin
Jun 13 2017 22:10:29.744Z DEBUG Setting user-agent to vic/v1.1.0-rc3
Jun 13 2017 22:10:29.784Z DEBUG URLFetcher.fetch() - &http.bodyEOFSignal{body:(*http.body)(0xc4201fbd00), mu:sync.Mutex{state:0, sema:0x0}, closed:false, rerr:error(nil), fn:(func(error) error)(0x6fee80), earlyCloseFn:(func() error)(0x6fee10)}, http.Header{"Server":[]string{"nginx/1.11.5"}, "Date":[]string{"Tue, 13 Jun 2017 22:10:30 GMT"}, "Content-Length":[]string{"527"}, "Docker-Content-Digest":[]string{"sha256:c79345819a6882c31b41bc771d9a94fc52872fa651b36771fbe0c8461d7ee558"}, "Docker-Distribution-Api-Version":[]string{"registry/2.0"}, "Etag":[]string{"\"sha256:c79345819a6882c31b41bc771d9a94fc52872fa651b36771fbe0c8461d7ee558\""}, "Content-Type":[]string{"application/vnd.docker.distribution.manifest.v2+json"}, "Connection":[]string{"keep-alive"}}
Jun 13 2017 22:10:29.784Z DEBUG [ END ] [github.com/vmware/vic/pkg/fetcher.(*URLFetcher).Fetch:124] [39.985863ms] https://192.168.31.208/v2/test/busybox/manifests/latest
Jun 13 2017 22:10:29.784Z DEBUG [ END ] [github.com/vmware/vic/lib/imagec.FetchImageManifest:313] [40.123683ms] 192.168.31.208/test/busybox:latest
```

**Comment by chengwang86** [comment]
Close it since it is a duplicate of https://github.com/vmware/vic/issues/5450

**Comment by chengwang86** [comment]
I suspect that this is because two concurrent builds were running the same test on the same ESXi host. The two Harbor instances had the same domain name `https://localhost.localdom.ci.vmware.run`. So when the VCH of the first build queried for its own Harbor ip, the dns-server returned the ip of the second Harbor's ip since the entry for the first Harbor's IP was overwritten.

**Comment by chengwang86** [comment]
Chatted with @mdharamadas1 about this issue. On the same test server, all the builds run sequentially. So there is no concurrent builds. 

This particular test passed in both ci and my local testing when `specific ci` is enabled. In my local testing environment, I had two harbor vms on the test server and the test still passed. 

Now we are wondering whether other tests might affect this particular test in full ci. I will run a full ci test to verify this.

Another possibility is that a harbor instance created in a previous build was not cleaned up appropriately and it affected the failed test.

**Comment by chengwang86** [comment]
Another reason might be:

In this test, we create two harbor instances, one with http and the other with https. After the main test is done, we only clean up the second harbor instance (govc vm.destroy) but leave the first harbor instance there. 

We should clean up both harbor instances right after they are not used.

**Comment by chengwang86** [comment]
Update:

Build 11504 seems to fail due to a different reason
```
Jun 13 2017 22:20:05.662Z ERROR Fetch (http://localhost.localdom.ci.vmware.run/service/token?scope=repository%3Atest%2Fbusybox%3Apull&service=token-service) to string error: Get https://localhost.localdom.ci.vmware.run/service/token?scope=repository%3Atest%2Fbusybox%3Apull&service=token-service: x509: certificate signed by unknown authority
```
It looks the same with #5622 and has been fixed by PR #5702 and #5756

**Comment by chengwang86** [comment]
I think we can close this ticket and restore the corresponding tests for now due to the following reasons:
- Based on my discussion with @mdharamadas1, this could be due to the leftover harbor instances that were not cleaned up appropriately from previous failed builds and the dns server might be messed up on the test server.
- This test passed in both my local test and ci test. The full ci test also passed after restoring these tests.

We can open it once we see it occurred again.
